base/linalg/triangular.jl

```diff
@@ -418,6 +418,8 @@ end
 scale!(A::Union{UpperTriangular,LowerTriangular}, c::Number) = scale!(A,A,c)
 scale!(c::Number, A::Union{UpperTriangular,LowerTriangular}) = scale!(A,c)
 
+fillstored!(A::AbstractTriangular, x) = (fill!(A.data, x); A)
```

Comment:
I wonder whether we should constrain the mutated part of `A.data` in case something else aliases `A.data`?

Reply:
It would be very weird to expect the following:
```julia
julia> A = fill(2., 2, 2);

julia> T = UpperTriangular(A);

julia> LinAlg.fillstored!(T, 0);

julia> A
2×2 Array{Float64,2}:
 0.0  0.0
 2.0  0.0
```
I think those wrappers should have the license to do whatever they like with the ignored part. (Also, the ignored part is still *stored* so the function name implies that we will fill that too :trollface: )

Reply:
The behavior you highlight does not strike me as weird, but rather as desirable :). Consider, for example, that packing the factors from a decomposition into a single matrix is common. Were you using `<:AbstractTriangular` wrappers to work with / manipulate those wrappers externally (which we do in `LinAlg` IIRC), the behavior above is correct, whereas the implemented behavior silently corrupts other objects. Another angle from which to consider this point: Should `fillstored!(view(A, 1:2, 1:2), 0)` corrupt all other data in `A`? :)

Reply:
A view clearly does not own it's wrapped Array while I would argue a type like `UpperTriangular` does.
We have no way in Julia to express these semantics but giving an Array to something like `UpperTriangular` and keep reusing that Array while having some expectations on how this type will modify it seems very suspicious and error prone.

Reply:
> A view clearly does not own it's wrapped Array

`<:AbstractTriangular` are also views, and just as a `SubArray` does not own its parent's storage beyond the `SubArray`'s purview, a `<:AbstractTriangular` should not either. Neither `SubArray`s nor `<:AbstractTriangular` allow mutation of the parent beyond the view's purview via `setindex!` and, insofar as I am aware, no other operations over such views allow/cause such mutation either. Implementations of operations over such views that do perform such mutation break the abstraction, and in this case unnecessarily so. Moreover, as mentioned above there are good reasons not to perform such mutation: Packed storage is common in numerical linear algebra, as in e.g. most dense triangular or orthogonal decompositions. See, e.g., base/linalg/lu.jl for operations over triangular views of subsections of packed decompositions that such mutation could jeopardize. Best!